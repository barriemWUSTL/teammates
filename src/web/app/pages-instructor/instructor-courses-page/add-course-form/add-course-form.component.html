<div class="card border-primary fill-plain">
  <div class="close-header">
    <button type="button" class="close" aria-label="Close" (click)="closeEditFormHandler()">
      <span aria-hidden="true">&times;</span>
    </button>
  </div>
  <div class="card-body text-center">
    <div class="form-group row">
      <label class="col-md-3 text-md-right col-form-label">Course ID:</label>
      <div class="col-md-6 text-md-left">
        <input id="new-course-id" class="form-control" type="text" [(ngModel)]="newCourseId"
            ngbTooltip="Enter the identifier of the course, e.g.CS3215-2013Semester1."
            [maxlength]="40" placeholder="e.g. CS3215-2013Semester1"/>
      </div>
    </div>
    <div class="form-group row">
      <label class="col-md-3 text-md-right col-form-label">Course Name:</label>
      <div class="col-md-6 text-md-left">
        <input id="new-course-name" class="form-control" type="text" [(ngModel)]="newCourseName"
            ngbTooltip="Enter the name of the course, e.g. Software Engineering."
            [maxlength]="64" placeholder="e.g. Software Engineering"/>
      </div>
    </div>
    <div class="form-group row">
      <label class="col-md-3 text-md-right col-form-label">Time Zone:</label>
      <div class="col-md-6 text-md-left">
        <div class="input-group">
          <select id="new-time-zone" class="form-control" [(ngModel)]="timezone" ngbTooltip="The time zone for the course. You should not
                need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
                to match the current time offset in your area, including clock changes due to daylight saving time.">
            <option *ngFor="let timezone of timezones" [value]="timezone">{{ timezone }}</option>
          </select>
          <span class="input-group-btn" style="margin-left: 5px;">
            <button class="btn btn-primary" (click)="onAutoDetectTimezone()">Auto Detect</button>
          </span>
        </div>
      </div>
    </div>
    <div class="form-group row">
      <div class="col-md-12">
        <button id="btn-save-course" class="btn btn-primary" (click)="onSubmit()">Add Course</button>
      </div>
    </div>
  </div>
</div>

<ng-template #newCourseMessageTemplate>
  The course has been added. Click <a routerLink="/web/instructor/courses/enroll" [queryParams]="{ courseid: course.courseId }">here</a>
  to add students to the course or click <a routerLink="/web/instructor/courses/edit" [queryParams]="{ courseid: course.courseId }">here</a>
  to add other instructors.<br>If you don't see the course in the list below,
  please refresh the page after a few moments.
</ng-template>
