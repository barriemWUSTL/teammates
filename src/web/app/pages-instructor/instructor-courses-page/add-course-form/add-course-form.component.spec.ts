import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { CourseService } from '../../../../services/course.service';
import { StatusMessageService } from '../../../../services/status-message.service';
import { TimezoneService } from '../../../../services/timezone.service';
import { Course } from '../../../../types/api-output';

import { HttpClientTestingModule } from '@angular/common/http/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';

import { of } from 'rxjs';
import { AddCourseFormComponent } from './add-course-form.component';

describe('AddCourseFormComponent', () => {
  let component: AddCourseFormComponent;
  let fixture: ComponentFixture<AddCourseFormComponent>;

  const testDate: Date = new Date('2020-13-05T08:15:30');
  const testCourseId: string = 'CS3281';
  const testCourseName: string = 'Valid course';
  const testTimeZone: string = 'UTC';
  const testCourse: Course =  {
    courseId: testCourseId,
    courseName: testCourseName,
    timeZone: testTimeZone,
    creationTimestamp: testDate.getTime(),
    deletionTimestamp: 0,
  };
  const timeZoneOffsets1: Record<string, number> = { GMT: 0 };

  const spyStatusMessageService: any = {
    showErrorToast: jest.fn(),
    showSuccessToastTemplate: jest.fn(),
  };
  const timezoneServiceStub: any = {
    getTzOffsets: jest.fn(() => timeZoneOffsets1),
    guessTimezone: jest.fn(() => 'UTC'),
  };
  const spyCourseService: any = {
    createCourse: jest.fn(() => of(testCourse)),
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [AddCourseFormComponent],
      imports: [
        HttpClientTestingModule,
        FormsModule,
        ReactiveFormsModule,
        RouterTestingModule,
        NgbModule,
      ],
      providers: [
        { provide: StatusMessageService, useValue: spyStatusMessageService },
        { provide: CourseService, useValue: spyCourseService },
        { provide: TimezoneService, useValue: timezoneServiceStub },
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(AddCourseFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    // Unable to leave timezone as default field, otherwise the field defaults to the
    // timezone the system is on. This will differ from
    // place to place causing the snapshot to constantly be mismatched.
    component.timezone = testTimeZone;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when not enabled', () => {
    component.isEnabled = false;
    component.timezone = testTimeZone;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should call showErrorToast when courseId is blank', () => {
    component.newCourseId = '';
    component.onSubmit();
    fixture.detectChanges();
    expect(spyStatusMessageService.showErrorToast).toHaveBeenCalled();
  });

  it('should hold added course with valid details', () => {
    component.newCourseId = testCourseId;
    component.newCourseName = testCourseName;
    component.onSubmit();
    fixture.detectChanges();
    expect(spyStatusMessageService.showSuccessToastTemplate).toHaveBeenCalled();
    expect(component.course).toEqual(testCourse);
  });
});
