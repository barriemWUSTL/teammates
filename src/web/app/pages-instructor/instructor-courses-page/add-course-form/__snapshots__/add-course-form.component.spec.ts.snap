// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCourseFormComponent should snap when not enabled 1`] = `
<tm-add-course-form
  closeCourseFormEvent={[Function EventEmitter]}
  courseAdded={[Function EventEmitter]}
  courseService={[Function Object]}
  isEnabled="false"
  newCourseId=""
  newCourseMessageTemplate={[Function TemplateRef_]}
  newCourseName=""
  statusMessageService={[Function Object]}
  timezone={[Function String]}
  timezoneService={[Function Object]}
  timezones={[Function Array]}
>
  <div
    class="card border-primary fill-plain"
  >
    <div
      class="close-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
      </button>
    </div>
    <div
      class="card-body text-center"
    >
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label"
        >
          Course ID:
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="new-course-id"
            maxlength="40"
            ngbtooltip="Enter the identifier of the course, e.g.CS3215-2013Semester1."
            placeholder="e.g. CS3215-2013Semester1"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label"
        >
          Course Name:
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="new-course-name"
            maxlength="64"
            ngbtooltip="Enter the name of the course, e.g. Software Engineering."
            placeholder="e.g. Software Engineering"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label"
        >
          Time Zone:
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <div
            class="input-group"
          >
            <select
              class="form-control ng-untouched ng-pristine ng-valid"
              id="new-time-zone"
              ngbtooltip="The time zone for the course. You should not
                  need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
                  to match the current time offset in your area, including clock changes due to daylight saving time."
            >
              
              <option
                value="GMT"
              >
                GMT
              </option>
            </select>
            <span
              class="input-group-btn"
              style="margin-left: 5px;"
            >
              <button
                class="btn btn-primary"
              >
                Auto Detect
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        class="form-group row"
      >
        <div
          class="col-md-12"
        >
          <button
            class="btn btn-primary"
            id="btn-save-course"
          >
            Add Course
          </button>
        </div>
      </div>
    </div>
  </div>
</tm-add-course-form>
`;

exports[`AddCourseFormComponent should snap with default fields 1`] = `
<tm-add-course-form
  closeCourseFormEvent={[Function EventEmitter]}
  courseAdded={[Function EventEmitter]}
  courseService={[Function Object]}
  isEnabled={[Function Boolean]}
  newCourseId=""
  newCourseMessageTemplate={[Function TemplateRef_]}
  newCourseName=""
  statusMessageService={[Function Object]}
  timezone={[Function String]}
  timezoneService={[Function Object]}
  timezones={[Function Array]}
>
  <div
    class="card border-primary fill-plain"
  >
    <div
      class="close-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
      </button>
    </div>
    <div
      class="card-body text-center"
    >
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label"
        >
          Course ID:
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="new-course-id"
            maxlength="40"
            ngbtooltip="Enter the identifier of the course, e.g.CS3215-2013Semester1."
            placeholder="e.g. CS3215-2013Semester1"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label"
        >
          Course Name:
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="new-course-name"
            maxlength="64"
            ngbtooltip="Enter the name of the course, e.g. Software Engineering."
            placeholder="e.g. Software Engineering"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label"
        >
          Time Zone:
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <div
            class="input-group"
          >
            <select
              class="form-control ng-untouched ng-pristine ng-valid"
              id="new-time-zone"
              ngbtooltip="The time zone for the course. You should not
                  need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
                  to match the current time offset in your area, including clock changes due to daylight saving time."
            >
              
              <option
                value="GMT"
              >
                GMT
              </option>
            </select>
            <span
              class="input-group-btn"
              style="margin-left: 5px;"
            >
              <button
                class="btn btn-primary"
              >
                Auto Detect
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        class="form-group row"
      >
        <div
          class="col-md-12"
        >
          <button
            class="btn btn-primary"
            id="btn-save-course"
          >
            Add Course
          </button>
        </div>
      </div>
    </div>
  </div>
</tm-add-course-form>
`;
